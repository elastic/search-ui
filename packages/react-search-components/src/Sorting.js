import PropTypes from "prop-types";
import React from "react";

function Sorting({ onChange, options, value }) {
  return (
    <div className="sui-sorting">
      <select
        id="sui-sorting__select"
        name="sorting"
        value={value}
        onChange={onChange}
      >
        {options.map(option => (
          <option key={option.value} value={option.value}>
            {option.name}
          </option>
        ))}
      </select>
    </div>
  );
}

Sorting.propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({ name: PropTypes.string, value: PropTypes.string })
  ).isRequired,
  value: PropTypes.string
};

export default Sorting;
