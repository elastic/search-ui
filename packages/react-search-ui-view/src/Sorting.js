import PropTypes from "prop-types";
import React from "react";
import Select from "react-select";

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({})
};

function Sorting({ label, onChange, options, value }) {
  const selectedValue = value;

  const selectedOption = selectedValue
    ? options.find(option => option.value === selectedValue)
    : null;

  return (
    <div className="sui-sorting">
      {label && <div className="sui-sorting__label">{label}</div>}
      <Select
        className="sui-select"
        classNamePrefix="sui-select"
        value={selectedOption}
        onChange={o => onChange(o.value)}
        options={options}
        isSearchable={false}
        styles={setDefaultStyle}
      />
    </div>
  );
}

Sorting.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({ value: PropTypes.string, label: PropTypes.string })
  ).isRequired,
  value: PropTypes.string
};

export default Sorting;
