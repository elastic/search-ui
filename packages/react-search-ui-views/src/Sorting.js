import PropTypes from "prop-types";
import React from "react";
import Select from "react-select";

import { appendClassName } from "./view-helpers";

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({})
};

function Sorting({ className, label, onChange, options, value, ...rest }) {
  const selectedValue = value;

  const selectedOption = selectedValue
    ? options.find(option => option.value === selectedValue)
    : null;

  return (
    <div className={appendClassName("sui-sorting", className)} {...rest}>
      {label && <div className="sui-sorting__label">{label}</div>}
      <Select
        className="sui-select"
        classNamePrefix="sui-select"
        value={selectedOption}
        onChange={o => onChange(o.value)}
        options={options}
        isSearchable={false}
        styles={setDefaultStyle}
      />
    </div>
  );
}

Sorting.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({ value: PropTypes.string, label: PropTypes.string })
  ).isRequired,
  value: PropTypes.string
};

export default Sorting;
