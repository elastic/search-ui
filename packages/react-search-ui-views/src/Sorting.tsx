import type { SearchContextState } from "@elastic/search-ui";
import React from "react";
import Select, { components } from "react-select";
import { BaseContainerProps } from "./types";

import { appendClassName } from "./view-helpers";

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({})
};

export type SortingContainerContext = Pick<
  SearchContextState,
  "sortDirection" | "sortField" | "sortList" | "setSort"
>;

export type SortingViewProps = BaseContainerProps &
  Pick<SortingContainerProps, "label"> & {
    onChange: (sortData?: any) => void;
    options: {
      value: string;
      label: string;
    }[];
    value: string;
  };

export type SortingContainerProps = BaseContainerProps &
  SortingContainerContext & {
    view?: React.ComponentType<SortingViewProps>;
    label?: string;
    sortOptions: any;
  };

function Option(props) {
  return <components.Option {...props}>{props.data.label}</components.Option>;
}

function Sorting({
  className,
  label,
  onChange,
  options,
  value,
  ...rest
}: SortingViewProps) {
  const selectedValue = value;

  const selectedOption = selectedValue
    ? options.find((option) => option.value === selectedValue)
    : null;

  return (
    <div className={appendClassName("sui-sorting", className)} {...rest}>
      {label && <div className="sui-sorting__label">{label}</div>}
      <Select
        className="sui-select"
        classNamePrefix="sui-select"
        value={selectedOption}
        onChange={(o) => onChange(o.value)}
        options={options}
        isSearchable={false}
        styles={setDefaultStyle}
        components={{
          Option: (props) => {
            props.innerProps["data-transaction-name"] = `sorting`;
            return Option(props);
          }
        }}
      />
    </div>
  );
}

export default Sorting;
