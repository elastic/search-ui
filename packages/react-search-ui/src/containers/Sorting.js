import PropTypes from "prop-types";
import { Component } from "react";
import { withSearch } from "..";
import { Sorting } from "@elastic/react-search-ui-views";
import SortList from "../types/SortList";

import { SortOption } from "../types";

function findSortOption(sortOptions, sortData) {
  if (sortData.indexOf("|||") === -1) {
    return sortOptions.find(
      option => JSON.stringify(option.value) === sortData
    );
  }

  const [value, direction] = sortData.split("|||");

  return sortOptions.find(
    option => option.value === value && option.direction === direction
  );
}

function formatValue(sortField, sortDirection, sortList) {
  if (sortList && sortList.length > 0) {
    return JSON.stringify(sortList);
  }
  return `${sortField}|||${sortDirection}`;
}

function formatSelectValues(sortOption) {
  if (Array.isArray(sortOption.value)) {
    // save value as string for comparison
    return JSON.stringify(sortOption.value);
  } else {
    return formatValue(sortOption.value, sortOption.direction);
  }
}

function formatSelectOption(sortOption) {
  return {
    label: sortOption.name,
    value: formatSelectValues(sortOption)
  };
}
export class SortingContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    label: PropTypes.string,
    sortOptions: PropTypes.arrayOf(SortOption).isRequired,
    view: PropTypes.func,
    // State
    sortDirection: PropTypes.oneOf(["asc", "desc", ""]),
    sortField: PropTypes.string,
    sortList: SortList,
    // Actions
    setSort: PropTypes.func.isRequired
  };

  render() {
    const {
      className,
      label,
      setSort,
      sortDirection,
      sortField,
      sortList,
      sortOptions,
      view,
      ...rest
    } = this.props;

    const View = view || Sorting;

    return View({
      className,
      label,
      onChange: o => {
        const sortOption = findSortOption(sortOptions, o);
        setSort(sortOption.value, sortOption.direction);
      },
      options: sortOptions.map(formatSelectOption),
      value: formatValue(sortField, sortDirection, sortList),
      ...rest
    });
  }
}

export default withSearch(
  ({ sortDirection, sortField, sortList, setSort }) => ({
    sortDirection,
    sortField,
    sortList,
    setSort
  })
)(SortingContainer);
