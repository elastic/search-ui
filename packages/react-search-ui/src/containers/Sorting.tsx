import React, { Component } from "react";
import { withSearch } from "..";
import {
  Sorting,
  SortingContainerContext,
  SortingContainerProps,
  SortingViewProps
} from "@elastic/react-search-ui-views";

function findSortOption(sortOptions, sortData) {
  if (sortData.indexOf("|||") === -1) {
    return sortOptions.find(
      (option) => JSON.stringify(option.value) === sortData
    );
  }

  const [value, direction] = sortData.split("|||");

  return sortOptions.find(
    (option) => option.value === value && option.direction === direction
  );
}

function formatValue(sortField, sortDirection, sortList) {
  if (sortList && sortList.length > 0) {
    return JSON.stringify(sortList);
  }
  return `${sortField}|||${sortDirection}`;
}

function formatSelectValues(sortOption) {
  if (Array.isArray(sortOption.value)) {
    // save value as string for comparison
    return JSON.stringify(sortOption.value);
  } else {
    return formatValue(sortOption.value, sortOption.direction, null);
  }
}

function formatSelectOption(sortOption) {
  return {
    label: sortOption.name,
    value: formatSelectValues(sortOption)
  };
}

export class SortingContainer extends Component<SortingContainerProps> {
  render() {
    const {
      className,
      label,
      setSort,
      sortDirection,
      sortField,
      sortList,
      sortOptions,
      view,
      ...rest
    } = this.props;

    const View = view || Sorting;

    const viewProps: SortingViewProps = {
      className,
      label,
      onChange: (o) => {
        const sortOption = findSortOption(sortOptions, o);
        setSort(sortOption.value, sortOption.direction);
      },
      options: sortOptions.map(formatSelectOption),
      value: formatValue(sortField, sortDirection, sortList),
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<SortingContainerProps, SortingContainerContext>(
  ({ sortDirection, sortField, sortList, setSort }) => ({
    sortDirection,
    sortField,
    sortList,
    setSort
  })
)(SortingContainer);
