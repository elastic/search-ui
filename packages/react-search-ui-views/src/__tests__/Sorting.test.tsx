import React from "react";
import Sorting from "../Sorting";
import { shallow } from "enzyme";
import { SortingViewProps } from "../Sorting";

const requiredProps: SortingViewProps = {
  onChange: jest.fn(),
  options: [
    { label: "Name ASC", value: "name|||asc" },
    { label: "Name DESC", value: "name|||desc" }
  ],
  value: "name|||asc"
};

it("renders correctly when there is a value", () => {
  const wrapper = shallow(<Sorting {...requiredProps} value={"name|||desc"} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders correctly when there is not a value", () => {
  const wrapper = shallow(<Sorting {...requiredProps} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <Sorting {...requiredProps} className={customClassName} />
  );
  const { className } = wrapper.props();
  expect(className).toEqual("sui-sorting test-class");
});
