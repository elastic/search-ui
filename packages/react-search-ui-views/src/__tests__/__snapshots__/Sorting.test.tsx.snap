// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there is a value 1`] = `
<div
  className="sui-sorting"
>
  <ForwardRef
    className="sui-select"
    classNamePrefix="sui-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "Name ASC",
          "value": "name|||asc",
        },
        Object {
          "label": "Name DESC",
          "value": "name|||desc",
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
      }
    }
    value={
      Object {
        "label": "Name DESC",
        "value": "name|||desc",
      }
    }
  />
</div>
`;

exports[`renders correctly when there is not a value 1`] = `
<div
  className="sui-sorting"
>
  <ForwardRef
    className="sui-select"
    classNamePrefix="sui-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "Name ASC",
          "value": "name|||asc",
        },
        Object {
          "label": "Name DESC",
          "value": "name|||desc",
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
      }
    }
    value={
      Object {
        "label": "Name ASC",
        "value": "name|||asc",
      }
    }
  />
</div>
`;
